docs: conceptual mapping of Keptn v1 components to KLT components (#1628)
keptn/lifecycle-toolkit: docs/content/en/docs/migrate/strategy/_index.md

We hope you will contribute your experiences
and questions that you have.

The Keptn Lifecycle Toolkit uses a different paradigm

> than Keptn v1 and so migration is not a straight-forward process.

Reviewer: this sentence sounds scary 😿 can we replace with something more positive? Something like "The Keptn Lifecycle Toolkit employs a distinct paradigm from Keptn v1, requiring technical adjustments during the migration process."

Author: done

Much of the logic and functionality of your Keptn v1 projects
can be migrated to KLT
but must be rewritten to utilize KLT components.

> ## How many namespaces?

Reviewer: I quite like this section! I am wondering if it should be the first one, probably as a v1 user I immediately care about project and services , than SLOs, and only at the very end, when I have an idea of what is happening, about how to allocate stuff in namespaces

Author: I put the namespaces stuff first because I was thinking this is something they need to address when they set up their cluster and before the install KLT.  But I'm thinking that most of this should really be moved to the general KLT installation section where we don't currently discuss namespaces at all.

> You have significan flexibility to decide how many namespaces to use

Reviewer: ```suggestion
You have significant flexibility to decide how many namespaces to use
```

### project

A Keptn v1 project represents an arbitrary, higher-level grouping of services.
A project is defined by a
[shipyard.yaml](https://keptn.sh/docs/1.0.x/reference/files/shipyard/)
file.

KLT does not recognize projects.
Some Keptn v1 projects may translate into `KeptnApp` resources
but many will not.
A project has a 1:1 mapping to a Git repo,
but, since few applications are stored in a mono-repo,
it is difficult to implement a project-as-application paradigm

In general, the Keptn v1 project is a useful as a reference list
when migrating to KLT
but does not directly translate to a KLT resource.

### service

A Keptn v1 service models a smaller chunk of a project.
Most projects include many services
In a micro-services world,
a service may represent a micro-service
but it could instead be a wrapper for something else,
like "the entire public website"

KLT does not have the concept of a service.
When migrating to KLT,
you need to analyze what each service is doing
and translate that into an appropriate resource.
The closest analogy is a Kubernetes
[orkload](https://kubernetes.io/docs/concepts/workloads/)
but some services may become
[Keptn tasks](../../implementing/tasks/)

> or other resources.

Reviewer: I would detail a bit this with examples, for instance, if a Keptn service was in charge of running chaos or load tests then they can probably be translated to a keptnTask using container runtime, if they used job executor they could also be a task. Instead, if they where a DB or a webserver etc they will be simple k8s workload resources, most likely respectively a stateful set and a deployment. For that, we can also refer to https://kubernetes.io/docs/concepts/workloads/controllers/

maybe others can have other examples too 😸

Author: Big improvement!  This is cut in.  Hopefully your examples will inspire others to come up with additional examples.

### stage

A stage is a subsection of a project
and has no corresponding KLT component.
However, the logic of the stages can be useful
when architecting the migration:

> * **deployment stage** -- KLT does not itself deploy the software

Reviewer: this sentence may be equivocal, in a sense KLT does deploy your app since it stops k8s and allows to bind pods to nodes...   I would maybe phrase it stating that: KLT does handle only the single application lifetime, it is not a continuous delivery tool since it does not provide promotion, thus for KLT there is no concept of stage. What it does is prevent the scheduling/deployment of an application version based on user-defined tasks/evaluations and integrate with other continuous delivery tools.

Author: I incorporated some of this info here and some at the beginning of the page where we discuss the main paradigm shifts for KLT versus Keptnv1.

    but rather depends on a deployment engine.
    However, a `deployment` stage may define sequences of tasks
    that should be translated into
    [KeptnTaskDefinition](../../yaml-crd-ref/taskdefinition.md)
    resources that are executed pre- and post-deployment
* **testing stage** may define sequences of tasks
    that should be translated into `KeptnTaskDefinition` resources

>     that are executed pre- and post-deployment.

Author: Users have these stages and need to figure out what to do with them.  I added your additional remarks; perhaps we can offer other examples and categorize stages that would be in different namespaces or clusters as we move forward.

    Note that all `KeptnTask` resources at the same level
    (either pre-deployment or post-deployment)
    execute in parallel
    whereas Keptn v1 sequences and tasks could execute in parallel.
    If you have actions that need to execute sequentially,
    create a single `KeptnTaskDefinition` that calls each action in order.
    If you have tasks that can execute in parallel,
    migrating to KLT may improve the performance of the deployment.

> ### sequence

Author: I moved this into the "Tasks" section and reworded it.  We have had requests to be able to execute tasks in parallel so I think this is actually a plus -- you can now choose what executes sequentially and what executes in parallel.

and (when the time is right),
emits a `taskName.triggered` cloud event onto the event bus.
The shipyard controller waits to receive a `taskName.started` event
and a correspondingly equal number of `taskName.finished` events
before the shipyard controller reads the next task
and emits a `taskName.finished` event for that task.

In this way, you can define arbitrary sequences of any tasks
at any length and also link (or chain) sequences together
to form (primitive) workflows.
When migrating, these sequences of tasks can often be translated into
[KeptnTaskDefinition](../../yaml-crd-ref/taskdefinition.md)
resources that are defined to run either pre- or post-deployment
of the pod-scheduling phase.

The `shipyard` file is a general purpose workflow engine
that is backed by cloud events.
It is not opinionated to a tool, platform, technology,
or a particular "slice" of the lifecycle.
The `TriggeredOn` property allows
a Keptn v1 sequence to be triggered at any time
by a user or another system.
This capability can be used, for example,
to trigger a data encryption and backup operation,
or a file movement over a network, or other arbitrary activities

> that may or may not have anything to do with an application's lifecycle.

Reviewer: should we shrink this down? maybe instead refer to v1 docs? I feel like this amount of text is unproportioned to the info we give for KLT and some v1 user may find it redundant maybe?

Author: Alas, the v1 docs are kind of weak in this area and I'm not going to fix them at this point.  I may, however, move much of this text to the "Paradigm changes" section at the beginning of this page.

Reviewer: sounds good, yeah I see no point into enriching v1 docs on this

When migrating to KLT,

> tasks that are not part of the lifecycle workflow

Reviewer: do you mean sequences?

Author: Yup -- fixed

should not be handled by KLT
but should instead be handled by the pipeline engine tools being used
such as Jenkins, Argo Workflows, Flux, and Tekton.

### task

> Keptn v1 defines some specific types of tasks,

Author: Yes.  And we actually have a page in this section to explain and illustrate this in more detail.

each of which is translated to KLT resources
that are appropriate for the activity:

* A **deployment task** becomes a
  [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
  workload.
  You can code
  [KeptnTaskDefinition](../../yaml-crd-ref/taskdefinition.md)
  and
  [KeptnEvaluationDefinition](../../yaml-crd-ref/evaluationdefinition.md)
  resources that are configured
  to run either pre- or post-deployment tasks

> * An **evaluation task** becomes a

Author: Yup, but we won't document that until we have it ;-)

  [KeptnEvaluationDefinition](../../yaml-crd-ref/evaluationdefinition.md)
  resource.
* All other standard tasks
  (**action**, **approval**, **get-action**, **rollback**,
  **release**, **test**)
  as well as custom task types
  that might be defined should be translated into
  `KeptnTaskDefinition` resources.

> * The `key:value` **properties** for each Keptn v1 sequence

Reviewer: also here there will be need to add stuff from analysis

  should be coded into the `KeptnTaskDefinition` resource
  as appropriate.

### SLIs

Keptn v1
[SLIs](https://keptn.sh/docs/1.0.x/reference/files/sli/)
(Service Level Indicators)
represent queries from the data provider
such as Prometheus, Dynatrace, or Datadog,
which is configured as a Keptn integration.

When migrating to KLT, you need to define a
[KeptnMetricsProvider](../../yaml-crd-ref/metricsprovider.md)
resource for the data provider you are using.

The queries defined for the Keptn v1 SLIs
should be translated into
[KeptnMetric](../../yaml-crd-ref/metric.md)
resources.
Note that KLT allows you to support multiple data providers
and multiple instances of each data provider for your SLI's
whereas Keptn v1 only allows you to use one SLI per project.

### SLOs

KLT at this time does not support the full range
of Quality Gates evaluations that are represented by
[SLOs](https://keptn.sh/docs/1.0.x/reference/files/slo/).
Facilities such as weighting of SLI's and scoring of the evaluation
do not currently exist.
However, simple evaluations of an SLI can be defined as
[KeptnEvaluationDefinition](../../yaml-crd-ref/evaluationdefinition.md)

> resources.

Reviewer: If we merge this soonish, shall we point to a coming soon 😸 ? https://github.com/keptn/lifecycle-toolkit/issues/1646

Author: Done.

### Remediation

KLT does not currently support the same level of
[remediations](https://keptn.sh/docs/1.0.x/reference/files/remediation/)
as Keptn v1 does.
KLT does provide limited "Day 2" facilities:

* Any query that is possible for your data provider post-deployment
  can be defined as a `KeptnMetricDefinition`;
  this is then reported as a Keptn Metric.
  Evaluation can be defined as a `KeptnEvaluationDefinition`.
* `KeptnMetricsDefinition` resources can be retrieved and used
  to implement the Kubernetes HorizontalPodAutoscaler (HPA),
  which can detect the need for additional resources
  (more pods, memory, disk space, etc)
  and automatically add those resources to your configuration

>   based on the `ReplicaSet` resources you have defined

Reviewer: maybe we could add that synce keptn clearly exposes metrics and crds related to each app, one could bring/design their own remediation ? Not sure really about what is there

Author: Maybe @thisthat can suggest appropriate remarks.  My understanding is that very few people are really using the remediation functionality of Keptnv1.  Mostly they want notification that something is wrong so they can go fix it.

follow the instructions in
[Installation and upgrade](../install)
to set up your Kubernetes cluster
and install KLT on it.

On this page, we discuss some of the major paradigm shifts
in KLT relative to Keptn v1
and then discuss how elements of your Keptn v1 can be implemented
for the Keptn Lifecycle Toolkit.

## The Keptn Lifecycle Toolkit paradigm

The Keptn Lifecycle Toolkit uses a different paradigm
than Keptn v1 and so migration requires technical adjustments.
Much of the logic and functionality of your Keptn v1 projects
can be migrated to KLT
but must be rewritten to utilize KLT components.

Some key points:

* KLT uses native Kubernetes CRDs
  to configure SLIs/SLOs, tasks, and other elements
  that are part of the environment where the application deployment lives
  rather than using its own Git repo and its
  [shipyard.yaml](https://keptn.sh/docs/1.0.x/reference/files/shipyard/)
  file as Keptn V1 does.
  - See the
    [CRD Reference](../../yaml-crd-ref)
    section for pages that describe the Keptn manifests
    that you populate manually for KLT.
  - See the
    [API Reference](../../crd-ref)
    for a comprehensive reference to all resources
    defined for Keptn.

> * KLT is not a continuous delivery tool
>   because it does not provide promotion.

Reviewer 2: ```suggestion
* KLT is not a delivery tool
  because it does not deploy applications.
```

>   Instead, it works with standard deployment tools

Reviewer 2: ```suggestion
  Instead, it focuses on supporting the deployment of your application in your Kubernetes environment.
  It works with standard deployment tools
```

  such as ArgoCD, Flux, even `kubectl -- apply`.
  KLT then prevents the scheduling and deployment
  of workflows if the environment does not meet
  the user-defined requirements.

* KLT operates on a
  [KeptnApp](../../yaml-crd-ref/app.md)
  resource
  that is an amalgamation of multiple Kubernetes workloads,
  which together comprise the released application.
  Each workload deploys a separate microservice,
  which can be deployed at different times using different tools.

* KLT integrates with continuous delivery tools
  to insures that a `KeptnApp` is not deployed
  if it does not meet the user-defined requirements
  for all the associated workloads.
  It also exposes metrics to evaluate the success/status of a deployment.

* KLT provides an operator
  that can observe and orchestrate application-aware workload life cycles.
  This operator leverages Kubernetes webhooks
  and extends the Kubernetes scheduler
  to support pre- and post-deployment hooks.
  When the operator detects a new version of a service
  (implemented as a Kubernetes
  [Workload](https://kubernetes.io/docs/concepts/workloads/)),
  it can execute pre- and post-deployment evaluations and tasks

>   using Kubernetes capabilities..

Reviewer 2: ```suggestion
  using Kubernetes capabilities.
```

* KLT provides extensive observability data
  using OpenTelementry and Prometheus
  rather than storing the data in a special Keptn database.
  This data can be displayed using Grafana and Jaeger
  or the dashboard of your choice.

For in-depth information about KLT components
and how they work, see the
[Architecture](../../concepts/architecture)
section.

## Disposition of Keptn v1 components in KLT

To help you wrap your mind around the migration process,
this lists Keptn v1 components
and identifies their possible disposition when you migrate to KLT.

### project

A Keptn v1 project represents an arbitrary, higher-level grouping of services.
A project is defined by a
[shipyard.yaml](https://keptn.sh/docs/1.0.x/reference/files/shipyard/)
file.

KLT does not recognize projects.
Some Keptn v1 projects may translate into `KeptnApp` resources
but many will not.
A project has a 1:1 mapping to a Git repo,
but, since few applications are stored in a mono-repo,

> it is difficult to implement a project-as-application paradigm

Reviewer 2: ```suggestion
it is difficult to implement a project-as-application paradigm.
```

In general, the Keptn v1 project is a useful as a reference list
when migrating to KLT
but does not directly translate to a KLT resource.

### service

A Keptn v1 service models a smaller chunk of a project.
Most projects include many services.
In a micro-services world,
a service may represent a micro-service
but it could instead be a wrapper for something else,
like "the entire public website"

KLT does not have the concept of a service.
When migrating to KLT,
you need to analyze what each service is doing
and translate that into an appropriate resource.
The closest analogy is a Kubernetes
[workload](https://kubernetes.io/docs/concepts/workloads/)
but some services may be translated into
[KeptnTaskDefinition](../../yaml-crd-ref/app.md)
or other resources.  See
[Working with Keptn tasks](../../implementing/tasks/)
for more information.

For example:

> * A Keptnv1 service that runs chaos or load tests

Reviewer 2: ```suggestion
* A Keptn v1 service that runs chaos or load tests
```

  can probably be translated into

>   a `KeptnTask` using the `container-runner`.

Reviewer 2: maybe here we can link the container-runner to its documentation page.

Author: It is already a link, actually ;)

> * A Keptnv1 service that runs a database

Reviewer 2: ```suggestion
* A Keptn v1 service that runs a database
```

  can probably be translated
  into a Kubenetes `StateFulSet` workload; see
  [Workload Resources](https://kubernetes.io/docs/concepts/workloads/controllers/)
  for more information.

> * A Keptnv1 service that runs a webserver

Reviewer 2: ```suggestion
* A Keptn v1 service that runs a webserver
```

  can probably be translated into
  a Kubernetes `Deployment` workload.

### stage

A stage is a subsection of a project.

> Because KLT is not a continuous delivery tool,

Reviewer 2: ```suggestion
Because KLT is not a delivery tool,
```
